Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,509 @@
+                        _ = http1PipelineConfigurator(channel.pipeline)
+                    }
+
+                    return protocolChoosenPromise.futureResult.flatMap { provider in
+                        self._connectionProviders[key] = provider
+                        return provider.getConnection(preference: preference)
+                    }
+                }
+            }
+        }
+    }
+
+    func release(_ connection: Connection) {
+        self.lock.withLock {
+            if let provider = self._connectionProviders[connection.key] {
+                provider.release(connection: connection)
+            } else {
+                fatalError("Attempting to release a connection that doesn't belong to the pool")
+            }
+        }
+    }
+
+    func closeAllConnections() -> EventLoopFuture<Void> {
+        return self.lock.withLock {
+            let closeFutures = _connectionProviders.values.map { $0.closeAllConnections() }
+            return EventLoopFuture<Void>.andAllSucceed(closeFutures, on: loopGroup.next())
+        }
+    }
+
+    /// Used by the `ConnectionPool` to index its `ConnectionProvider`s
+    ///
+    /// A key is initialized from a `URL`, extracting, validating and storing its scheme,
+    /// host and port components. It uses the components to derive a hashed value
+    /// used by the `connectionProviders` dictionary to allow retreiving and
+    /// creating connection providers associated to a certain request in constant time.
+    struct Key: Hashable {
+        init(url: URL) throws {
+            switch url.scheme {
+            case "http":
+                self.scheme = .http
+            case "https":
+                self.scheme = .https
+            default:
+                if let scheme = url.scheme {
+                    throw HTTPClientError.unsupportedScheme(scheme)
+                } else {
+                    throw HTTPClientError.emptyScheme
+                }
+            }
+
+            switch url.port {
+            case .some(let specifiedPort):
+                self.port = specifiedPort
+            case .none:
+                switch self.scheme {
+                case .http:
+                    self.port = 80
+                case .https:
+                    self.port = 443
+                }
+            }
+
+            guard let host = url.host else {
+                throw HTTPClientError.emptyHost
+            }
+            self.host = host
+        }
+
+        var scheme: Scheme
+        var host: String
+        var port: Int
+
+        enum Scheme: Hashable {
+            case http
+            case https
+        }
+    }
+
+    class Connection {
+        init(key: Key, channel: Channel, parentPool: ConnectionPool) {
+            self.key = key
+            self.channel = channel
+            self.parentPool = parentPool
+        }
+
+        func release() {
+            self.parentPool.release(self)
+        }
+
+        unowned let parentPool: ConnectionPool
```

Comment:
The use of `unowned let` here is probably an anti-pattern: it suggests that a `Connection` is a peer of a `ConnectionPool`, which it isn't really: it's strictly an owned object. Is there a reason we need to be able to call upwards? Shouldn't this just be called on the parent pool directly?

Reply:
Having the pool as a stored property allows uses [like this one](https://github.com/swift-server/async-http-client/pull/105/files#diff-67e2f60fc709da32af89dfbfc0c774aaR449), I could also add a reference to the connection pool indeed, but I thought it could be nice to just call `.release()` on a `Connection`?

About `unowned` I'm not sure to understand what you mean when you say that it suggests that `Connection` is a peer of `ConnectionPool`? I would have thought `weak` would have suggested this while `unowned` means the `Connection` is the child of the `ConnectionPool` and does not own it?

Reply:
To explain a little better: the reason why I wanted to enable calling `.release()` on a `Connection` was to make it syntactically similar (and as simple) to how `.close(...)` is called on `channel` with the current implementation (in master). That's because with the proposed changes, the `Channel` is now itself wrapped in `Connection`.